impl/src/main/java/com/okta/sdk/impl/http/authc/OAuth2RequestAuthenticator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.http.authc;
+
+import com.okta.commons.http.Request;
+import com.okta.commons.http.authc.RequestAuthenticationException;
+import com.okta.commons.http.authc.RequestAuthenticator;
+import com.okta.sdk.authc.credentials.ClientCredentials;
+import com.okta.sdk.impl.oauth2.OAuth2AccessToken;
+import com.okta.sdk.impl.oauth2.OAuth2ClientCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+
+/**
+ * This implementation used by OAuth2 flow adds Bearer header with access token as the
+ * value in all outgoing requests. This has logic to fetch a new token and store it in
+ * {@link OAuth2ClientCredentials} object if the existing one has expired.
+ *
+ * @since 0.6.0
+ */
+public class OAuth2RequestAuthenticator implements RequestAuthenticator {
+    private static final Logger log = LoggerFactory.getLogger(OAuth2RequestAuthenticator.class);
+
+    private final ClientCredentials<OAuth2AccessToken> clientCredentials;
+
+    public OAuth2RequestAuthenticator(ClientCredentials<OAuth2AccessToken> clientCredentials) {
+        this.clientCredentials = clientCredentials;
+    }
+
+    @Override
+    public void authenticate(Request request) throws RequestAuthenticationException {
+
+        if (clientCredentials != null &&
```

Comment:
This should never be null, since it's `final` and passed in via the constructor, maybe just assert not null there?

Reply:
The client credentials (oauth2) is used by Token Client & API Client. So, when the client builder creates the credentials resolver to be used by token client initially, the client credentials would be null. The code snippet in question will ensure we do not take any action (add Bearer header) for such token requests. Now, the later part in DefaultClientBuilder would set the client credentials resolver object with accessTokenRetriever service instance. This object will contain the eager fetched token and ensure that that authenticator code in line #75 will set the Bearer token. The bottom line is, client credentials not being checked for null in the constructor here is intentional.

Reply:
I think I'm missing something. Any time when the `clientCredentials` is null in the constructor this `RequestAuthenticator` would basically be a no-op, which I would assume is an error case.  If you really need a no-op use the `DisabledAuthenticator` instead.

Reply:
Refactored code around this. Please take a look.